LC Course Prototype

kdmccormick opened the pull request:
Uses https://github.com/openedx/openedx-learning/pull/326

Commit:
feat: redesign import task + api/helpers tests adaptation

Commit:
feat: add simple REST API

Commit:
refactor: add API endpoint secs to view docs, updated tests & fix various lint errors

Commit:
refactor: simplify staged content in helpers tests & fix import order in tasks.py

Commit:
fix: add default value for composition_level to not break old instances

kdmccormick committed:
feat!: modulestore_migrator (WIP)

kdmccormick committed:
feat: add admin interface

kdmccormick committed:
feat: better task debug info on the admin interface

kdmccormick committed:
fix: fixes to tasks and admin -- happy path works now

kdmccormick committed:
temp: install openedx-learning@lc-course-prototype

kdmccormick committed:
Merge branch 'kdmccormick/import-from-modulestore' into lc-course-prototype

kdmccormick committed:
fix: install outline_roots and courses from LC

kdmccormick committed:
feat: Migrate an OutlineRoot into a library

kdmccormick commented:
WIP:
* https://github.com/openedx/edx-platform/pull/36913

kdmccormick committed:
fix: Create ContainerVersion subclassses too

This commit stops this error:

   ContainerVersion.unitversion.RelatedObjectDoesNotExist: ContainerVersion has no unitversion.

from happening when accessing migrated content.

You will have to re-migrate the content to fix. You may even have to discard the
original broken migrated content.

ormsbee committed:
feat: hacky splitmodulestore shim writing from learning core

kdmccormick committed:
Merge pull request #36913 from openedx/lc-course-prototype-migrate-outline

feat: Migrate an OutlineRoot into a library

ormsbee committed:
temp: more debugging

Agrendalath committed:
feat: store XBlock field data in MySQL

Agrendalath committed:
fixup! feat: store XBlock field data in MySQL

kdmccormick committed:
fix: drop all the models to un-bork modulestoreblocksource

Agrendalath committed:
fixup! feat: store XBlock field data in MySQL

kdmccormick committed:
fix: recreate models with missing db constraint

we were missing the unique_together on modulestoreblocksource,
meaning we had multiple modulestoreblocksources for the
same legacy usage key.

also, replace bulk_create with a series of saves to un-bork
forwarding for now.

also, this renames forwarded_by -> forwarded

ormsbee committed:
temp: disable shim

ormsbee committed:
temp: writing parent/child relations for import works

ormsbee committed:
temp: the most basic editing in library works, but you still have to rebuild the structure doc manually

ormsbee committed:
temp: switch LC vs. ModuleStore based on lookup table

kdmccormick committed:
feat: Migrate to CourseRun, including Meta-Components

ormsbee committed:
temp: LC course publishing from library

ormsbee committed:
temp: auto-add new content to containers, disable outline caching

ormsbee committed:
temp: debug logging

ormsbee committed:
temp: no longer need to fake the non-attached blocks

ormsbee committed:
refactor: move more xblock-data logic into xblock app

ormsbee committed:
feat: merge pull request #36938 from ormsbee/lc/modulestore-write

Prototype Writing to SplitModuleStore-shim layer from Learning Core Libraries

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10453706705.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10453706705-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10453706705-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10453706705-tutor-requirements.txt)

kdmccormick committed:
Merge branch 'master' into lc-course-prototype

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10467295658.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10467295658-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10467295658-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10467295658-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10467295662.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10467295662-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10467295662-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10467295662-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-36909-139931.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-36909-139931.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10473960464-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10473960464-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10473960464-tutor-requirements.txt)

ormsbee committed:
fix: don't add stuff to the structure if we don't have field data for it

Also special case ignore "discussion" blocks, for those libraries where
we already wrote that data into the structures in places.

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-36909-139931.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-36909-139931.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10506022477-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10506022477-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36909-139931-51564656-10506022477-tutor-requirements.txt)

bradenmacdonald commented:
For the record, here's how I set up the sandbox for this:

1. Import [the demo course](https://github.com/openedx/openedx-demo-course/)
2. Create an empty library
3. Go to https://studio.pr-36909-139931.sandboxes.opencraft.hosting/admin/modulestore_migrator/modulestoresource/add/
   Add source with key `course-v1:...` from step 1.
4. Go to https://studio.pr-36909-139931.sandboxes.opencraft.hosting/admin/modulestore_migrator/modulestoresource/
   Action: Start migration (select the course from step 3), Target: library key from step 2 (e.g. `lib:OpenedX:Demo`), Replace: YES, Aggregate up to CourseRun, Go
5. Go to the library and publish all.
6. Repeat steps 4 and 5 (workaround required for issues in the prototype)
7. Verify at https://studio.pr-36909-139931.sandboxes.opencraft.hosting/admin/xblock_new/learningcorelearningcontext/ that the course is now enabled for Learning Core
8. Go to https://app.pr-36909-139931.sandboxes.opencraft.hosting/learner-dashboard/ and view the course!

kdmccormick commented:
Talk slides are here: https://docs.google.com/presentation/d/1jxl67EC7dyo3NMXf2yk_eZKLLtBEoh7LvtsbXKnxxXs/edit?slide=id.g36961410b1f_0_8#slide=id.g36961410b1f_0_8

Talk video including a demo of this PR in action will be uploaded soon to https://www.youtube.com/@OpenedX
